### ISSUE TITLE
development: Add development containers to repos and templates

### ISSUE THREAD

**Issue by christopher-davis-afs** [EVerest/EVerest-archived#120]
**Is your feature request related to a problem? Please describe.**
When developing projects in EVerest, my team found it difficult to work on individual projects and setting up all of the dependencies without using the workspace set up by `everest-utils`.

**Describe the solution you'd like**
A solution to help developers get set up and working on EVerest repos would be to have `.devcontainer` folders with a `devcontainer.json` in each repo. This `devcontainer.json` file would reference a docker image with the standard development tools required to build and test EVerest repos. This should be kept in sync with the dependencies installed in the EVerest `build-kit` images, but use the standard development container images as their base instead of plain docker images.

An example of this can be seen at https://github.com/US-JOET/everest-template/tree/wip/cdavis/add-dev-container/.devcontainer

This enables developers using tools like Visual Studio Code to quickly stand up a development environment that is fully ready to build and test the application.

My team thinks that this would work best if this was also included in the everest-template repo so that it is carried forward in new repositories.

The primary question: Where should the dockerfile and setup to publish this docker image live?

**Comment by Dominik-K** [comment]
Nice idea to use `.devcontainer` to simplify the definition of the workspace and the aim to get it working out of the box. It seems this approach is getting attraction already outside of VSCode, e.g. at [IntelliJ IDEs](https://blog.jetbrains.com/idea/2023/06/intellij-idea-2023-2-eap-6/#support-for-dev-containers). :+1: 

However, I dislike that the Dockerfile is basically a copy of the `build-kit`-image. Isn't there a way to directly reference to that one instead? It's public, so shouldn't be a problem to use this.

I'm unsure if we should to add the `customizations` in there. It feels like this would become quite a lot boilerplate if we would allow this for all kinds of different IDEs. If it's only bound to VSCode, then it's simply a replacement for `.vscode` which we (intentionally?) left out in nearly all repos. But I'm open to discuss this.

One thing to consider is that we might define the `.devcontainer`-directory once for all repos and add it via Git submodules. To ease the maintenance of it. But that relies on how many stuff we put in there, IMO.

**Comment by christopher-davis-afs** [comment]
Thanks for your feedback :) To address your comments:

> ...I dislike that the Dockerfile is basically a copy of the build-kit-image. Isn't there a way to directly reference to that one instead? It's public, so shouldn't be a problem to use this.

`build-kit` does not come with all of the required tooling to use the images as a proper development container. We would need to essentially rebuild the base development container setup on top of `build-kit` instead of using that as a base. Since `build-kit` is relatively small I think it's better to use the devcontainers base and put what we need on top.

> I'm unsure if we should to add the customizations in there. It feels like this would become quite a lot boilerplate if we would allow this for all kinds of different IDEs. If it's only bound to VSCode, then it's simply a replacement for .vscode which we (intentionally?) left out in nearly all repos. But I'm open to discuss this.

This is a good point. I personally think it's better for developers if the `customizations` are there, as they prepare everything a developer (using VS Code at least) would need to get started in their editor without any extra steps. All they need to do is clone the repo, open it in VS Code, and click the prompt to open the container and they'll be ready to develop.

> One thing to consider is that we might define the .devcontainer-directory once for all repos and add it via Git submodules. To ease the maintenance of it. But that relies on how many stuff we put in there, IMO.

If we get the `dev-kit` docker image published, most repos would only need the `devcontainer.json` file, so I think a submodule is a bit much for that.

**Comment by ghost** [comment]
@andistorm notifying you based on the community call to review.

**Comment by christopher-davis-afs** [comment]
I created a PR against `everest-template` with a sample container. The goal is for developers to have get development tools and helpful VSCode extensions installed and properly configured out of the box. Please let me know if I'm missing anything required for working on Everest projects in here:

https://github.com/EVerest/everest-template/pull/2

**Comment by andistorm** [comment]
I like the idea of unifying the local dev enviroments. I also like to use devcontainer for that.
A few thoughts on that:

1. I didn't fully understand the issue with the already defined devcontainer in `everest-utils`. Could you explain what exactly the issue is here?
2. It is hard to maintain devcontainer for each repository. 
3. Since devcontainer or at least parts of the devcontainer.json are specific for VSCode I would not put it next to the code
4. In case of using the build-kit image as base, what would be needed to install on top of it?

In my opinion we should find a way to only define one Dockerfile and one `devcontainer.json` which is usable for all repos.
This could be defined as now in `everest-utils`. Maybe it would fit better into everest-dev-environmet (t.b.d).
I could imagine to support multiple devcontainers in various falvours, as done in `everest-utils`:
- https://github.com/EVerest/everest-utils/tree/main/docker/everest-playground-opensuse-tumbleweed
- https://github.com/EVerest/everest-utils/tree/main/docker/everest-playground-ubuntu-20-04
- https://github.com/EVerest/everest-utils/tree/main/docker/everest-playground

Since I currently don't have many resources available, I would appreciate it if you could suggest a design that addresses the points mentioned above.

**Comment by ghost** [comment]
1. Currently the devcontainer in `everest-utils` setups all of everest. It makes development on a specific library like libocpp much more complicated then needed. The goal of a devcontainer is to only run what is needed for the specific repo's code that is being worked on.
    - EDM is nice tool that is plugged into a lot of what this devcontainer does but the documentation is limited on it.
    - It was difficult to have a forked branch and different version of all of everest running
    - Generally testing and debugging was much more bogged down in the larger devcontainer.
2. We have discussed the current implementation of the devcontainer. One item is to have the docker image built in the CI repo and then referenced per a repo which would remove the Dockerfile from each repo. That way there is only one place the build occurs. Having the devcontainer (optionally the Dockerfile and json) in each repo gives repos the flexibility to customize what is needed for that specific unit of work. Generally speaking it is better to have it set per repo. If you could specify what maintenance issues you are seeing we could come up with some solutions.
3. VSCode is probably the most active usecase of devcontainers but the json supports many other IDEs. Since the devcontainer and the IDE is setup for the specific code the files must live in the reop of the code they are supporting.
5. The build-kit is either alpine of slim which are not particularly useful for development work. The shell is significantly modified and some of what is shared between the host and virtual is helpful ssh and gpg particularity.

I would ask not to create another repo. It will make it harder to find over all and isolate it from where it is being used. As for various flavors of OS that is not really the purpose of a shared development environment. We want to empower the community to use this tool to develop consistently and effectively. The CI/CD pipeline is where you would want to offer this variability with multiple build profiles to tests across those systems.

**Comment by andistorm** [comment]
Thanks for your answer. I think I'm getting closer to fully understand here :)
I want to try to answer in a structured way:

### 1. Advantages of devcontainers
Let me summarize why we discuss this topic at all.
There are several advantages that come with devcontainers:
#### 1.1. No need to worry about dependencies / requirements at all
They are defined by the Docker image. In case of breaking one developer can fix it for all.
#### 1.2. No conflicting dependency / requirement versions on host system
Sometimes it is necessary to use a specific version of something in a `project`. In case of using this package already on host, it can come to conflicts between versions.
#### 1.3. Allowing multiple different versions of packages across `projects`
Since inside a `project` it shouldn't be necessary having different versions of one package (in the best case), a developer maybe works on multiple `projects` that each require another version.
#### 1.4. Being able to run github codespace based on devcontainer
A devcontainer allows using the github codespace feature, which could be useful to use in case of not having access to the normally used hardware for example.
#### 1.5. Dev environments become unified, errors become more reproducable
Since the dependency / requirement versions should be pinned to a specific version, the devcontainer should make it easier to reproduce errors occuring on anothers developer environment.
#### 1.6. Set of dependencies / requirements can be restricted to what is needed for the `project`
With devcontainers it is also possible to "restrict" the dependencies / requirements to an well defined set of packages.

### 2. Possible ways to work on the dev environemt
I agree on that the current situation is not ideal and the developer experience could be better.
There are a couple ways we could go to improve that. In the end I think it is, at least partially, a matter of taste.
#### 2.1. Having only one `devcontainer.json` and one `Dockerfile` for all EVerest repositories

Advantages:
- One place to work on the dev environment
- Easy to switch repositories you're developing on
- Easy to work on multiple repositories at the same time
- You always know what is included in the current devcontainer / It is easy to understand 

Disadvantages:
- Need to be documented well, to allow new contributors easy acess / entrypoint
- No individual settings for single repositories
- There might be VSCode extensions (or other packages) that are not needed for the specific thing a developer is working on

#### 2.2. Having one `devcontainer.json` for each repository that refers to a common `Dockerfile` / docker image

Advantages:
- Only one `Dockerfile` that can be deployed as one docker image at ghcr.io
- One place for host requirements
- Easy entrypoint / Easy to use for new contributors even without documentation

Disadvantages:
- Even if there is only one docker image there are multiple versions of this image that doesn't need to align on multiple `devcontainer.json` files
- Might be confusing in case of features that are available in devcontainer A but not in devcontainer B
- You would need to open multiple VSCode instances to work on multiple repositories in parallel
- Introducing new features / dependencies / requirements need to be done at multiple locations, in worst case multiple developers are doing the same thing at the same time...

#### 2.3. Having one `devcontainer.json` and one `Dockerfile` for each repository in EVerest
The `Dockerfile`s could share one common base image. But don't need to and can add/remove things individually.

Advantages:
- Always being restricted to necessary stuff
- Easy entrypoint / Easy to use for new contributors even without documentation

Disadvantages:

The disadvantages are the same as `2.2.`, with the exception of having the `Dockerfiles` aligned with the `devcontainer.json` files

### 3. Definitions
#### 3.1. `project`
In my view a project can be a single repository, but it could also be a set of multiple repositories depending on each other.

### 4. My opinion.

#### 4.1 My personal definition of `project`
In our case "EVerest" I would define `project` as a set of multiple repositories (located in the github org EVerest).
Daily work happens on multiple repositories and issues trigger changes at more than one repository.

#### 4.2 How my daily dev environemt looks like
I have a directory called `~/EVworkspaces/` in my home directory. If i pick up an issue for development I create a new subdirectory named after the issue. In this directory I clone a bunch of repositories I need to work on for this issue.
In the development process it regulary happens that I quickly add a repository in such a workspace. This is why I also create a workspace when I expect to only work on one repository (Things don't come as expected :) ).

In VSCode I always open the complete workspace directory, this way I am able to have one VSCode instance per issue/task I am working on.

At the moment I don't really work with devcontainers. But I see the advantages of that and plan to try out using them.

#### 4.3 Which option I would prefer
Having a look at the options under point `2` it shows that each option has its advantages and disadvantages at the end it is a matter of taste or better said a weighting of these advantages/disadvantages.

As I already mentioned in the post before I tend to option `2.1. Having only one `devcontainer.json` and one `Dockerfile` for all EVerest repositories`. A discussion today with @hikinggrass showed that we align here.

In my opinion the definition of what a `project` is, is very important. With my defintion all advantages in `1. Advantages of devcontainers` are covered.
As I understood you see a `project` more as a repository, so I can understand why you might say not all advantages are included with option `2.1.`

#### 4.4. How to solve/invalidate the disadvantages
Currently we are working on an improved documentation of EVerest that will be well structured and will allow easy access for new contributors, so it should be managable to find documentation for devcontainers easy in future.
I would say not having individual devcontainers for each repo can be an advantage too, because this way you have good chances to get your devcontainer known well. And it's the same for any issue / task you're working on. You won't miss any feature, but if you do so you can add this feature to all (the one) devcontainer and all developers can profit from this.

### 5. Facit
I would say we shouldn't focus on things like the used base image in this discussion at the moment, since such things are implementation details. Let's agree on a general design first.

Sorry, if I overwhelmed you a bit with the long response.

Please let me know if there are points I made that are not clear to you, if I forgot important advantages / disadvantages / options or if you totally disagree with something.

**Comment by ghost** [comment]
Thank you very much for the detailed response. It is genuinely helpful. I find your outline of the reasons for the discussion and the pros/cons to be quite accurate. As with every tool in software development, there are trade-offs. The worst-case scenario is letting perfect be the enemy of good. No solution will address all issues. Our aim is to establish the most standard and common use case for devcontainers, which is on a per-repository basis. The idea is to facilitate the quick onboarding of eager contributors and minimize maintenance for the community as a whole.

I don't dispute how you and your team define a project. If that workflow suits your setup and is effective for you, then it's the right approach. From an external perspective, my team considers a repository a project because it is an independent dependency. Each repo/dependency should be decoupled from the code that uses it. For example, to work on libocpp, I do not need to have libevse-security locally. I may find a need to update it, but that should be the exception.

However, based on the description of your workflow, I understand why having devcontainers per repository may be a concern for you. Nevertheless, you can set up one VSCode instance for multiple remote devcontainers [https://code.visualstudio.com/remote/advancedcontainers/connect-multiple-containers](https://usg02.safelinks.protection.office365.us/?url=https%3A%2F%2Fcode.visualstudio.com%2Fremote%2Fadvancedcontainers%2Fconnect-multiple-containers&data=05%7C02%7Ccoury.richards%40afs.com%7C1bd8a38fe0694da4a6ea08dc2f3e545e%7Ca01f407a85cb4a1698bbf28e6384bd28%7C0%7C0%7C638437191951747741%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=cx1QmwOODrms5ECz6%2FIbSPmertzAhU3PUAmmX8gXZM4%3D&reserved=0). Although I haven't personally done this, it is an option for your workflow.

Our current workflow relies more heavily on git branches and the use of EDM over additional subdirectories.

Our basic workflow looks something like this:

1. Fork the necessary Everest repos to our org.

2. On disk, we pull down to a location like ~/Documents/code/us-joet/.

3. In that directory, we can have multiple repos, similar to what you have in EVWorkspace.

In the simplest case, we are adding features to one repo (i.e., libocpp).

1. Create a branch for our current work.

2. Open VSCode connected to a devcontainer.

3. Make the needed changes.

4. Add additional tests.

5. Build and test locally.

6. Push these changes to remote.

7. Verify CI.

The more complex version is if we need to go across repos (i.e., libocpp and libevse-security). Much of the same workflow above applies:

1-7 above for changes in libevse-security.

8. Update EDM in libocpp to point to the remote branch in our org.

9. Rinse and repeat 1-7 for libocpp.

10. If we need to go back and update `libevse-security`, we can and push to remote.

11. From here, we would have to make sure to update EDM before pushing to Everest.

12. Ensure to keep the dependent PR in draft and reference the PR that should be merged first.

This is just our current workflow. Many other workflows can occur, and devcontainers support a variety of cases. The reason for this proposal is to assist new developers with narrow goals. It's challenging to expect them to find the correct documentation in a location separate from the code they are working on. I also don't see anything wrong with completely ignoring the devcontainers if they do not serve your needs.

**Comment by andistorm** [comment]
Thank you for your patience on this matter.

We can now make our stance clear: We do not wish to have your proposal implemented in its current form. The vision here is to initially have only one fully-featured devcontainer (.devcontainer.json + Dockerfile) that can be used for as many repositories as possible. The reason for this is the preference not to view individual repositories as independent projects and to avoid the extra effort in maintaining these files.

Ignoring these additional files is unfortunately not a solution from our perspective, as it may lead to situations in the future where support requests arise, and it would be difficult to justify having "unsupported" features. In other words, we want to offer only what we can maintain/support. Technically, I understand your approach and would generally consider it a possible solution, but at this point, it's probably a matter of personal preference.

I would also like to thank you for the discussion, as it has raised the issue in the first place. Even though we are not implementing the proposal made here, we have definitely made a good step forward.

Now, I would like to delve a bit more into how we envision the solution. As mentioned earlier, we would start with just one docker image and only one .devcontainer.json file. However, the usage should be as convenient as possible, especially in VSCode.
VSCode configures various paths where such configurations are searched for. So, it is conceivable that you only need to place the devcontainer once in the right location.
The idea is to use the command line tool `edm`, so that once to set up the development environment, you would call a command like `edm setup-devcontainer`. Behind such a command should be the magic of making the devcontainer as accessible as possible.

This decision is so far only a basic design decision on this matter. If you are interested in working with us in this direction, we look forward to discussing further details (such as which Base Image to choose). Sooner or later, we would also develop this solution ourselves, but at the moment, the capacity is simply lacking, and we would push it onto our backlog.
